sparse/densification.py

```diff
@@ -0,0 +1,140 @@
+import sys
+from numbers import Integral, Number
+from collections import Iterable
+
+import numpy as np
+
+from .utils import _get_broadcast_shape, SparseArray
+
+try:  # Windows compatibility
+    int = long
+except NameError:
+    pass
+
+try:
+    _max_size = sys.maxsize
+except NameError:
+    _max_size = sys.maxint
+
+
+class DensificationConfig(object):
+    def __init__(self, densify=None, max_size=None, min_density=None):
+        if isinstance(densify, (Iterable, DensificationConfig)):
+            self.make_copy_of(densify)
+            return
+
+        if max_size is None:
+            max_size = 10000
+
+        if min_density is None:
+            min_density = 0.25
+
+        if not isinstance(densify, bool) and densify is not None:
+            raise ValueError('always_densify must be a bool or None.')
+
+        if not isinstance(max_size, Integral) or max_size < 0:
+            raise ValueError("max_nnz must be a non-negative integer.")
+
+        if not isinstance(min_density, Number) or not (0.0 <= min_density <= 1.0):
+            raise ValueError('min_density must be a number between 0 and 1.')
+
+        self.densify = densify
+        self.max_size = int(max_size)
+        self.min_density = float(min_density)
+
+    def make_copy_of(self, other):
+        if isinstance(other, Iterable):
+            single = DensificationConfig.from_many(other)
+        else:
+            single = other
+
+        self.densify = single.densify
+        self.max_size = single.max_size
+        self.min_density = single.min_density
+
+    def _should_densify(self, size, density):
+        if self.densify:
+            return True
+        elif self.densify is False:
+            return False
+        else:
+            return self.max_size >= size or self.min_density <= density
+
+    def _raise_if_fails(self, size, density, name):
+        if not self._should_densify(size, density):
+            raise ValueError("Performing this operation would produce "
+                             "a dense result: %s" % name)
+
+    def check(self, name, *arrays):
```

Comment:
`name` could be made optional. varargs like `*arrays` can become a huge PITA, I think people should pass an iterable instead.

Reply:
Name is needed, we use it for the exception.

Reply:
I'm willing to change this, but it is a huge ease to simply do `config.check(name, arr)` or `config.check(name, arr1, arr2)` instead of putting everything into a list. If we need to pass a list, we can always do `config.check(name, *arrays)`. Could you tell me why it's difficult to support?